modules/platforms/dotnet/Apache.Ignite.Core/Impl/Client/Transactions/TransactionClient.cs

```
@@ -0,0 +1,172 @@
+            TransactionIsolation isolation,
+            TimeSpan timeout,
+            string label)
+        {
+            _id = id;
+            _ignite = ignite;
+            _socket = socket;
+            Concurrency = concurrency;
+            Isolation = isolation;
+            Timeout = timeout;
+            Label = label;
+        }
+
+        /** <inheritdoc /> */
+        public void Commit()
+        {
+            ThrowIfClosed();
+            Close(true);
+        }
+
+        /** <inheritdoc /> */
+        public void Rollback()
+        {
+            ThrowIfClosed();
+            Close(false);
+        }
+
+        /** <inheritdoc /> */
+        public TransactionConcurrency Concurrency { get; private set; }
+
+        /** <inheritdoc /> */
+        public TransactionIsolation Isolation { get; private set; }
+
+        /** <inheritdoc /> */
+        public TimeSpan Timeout { get; private set; }
+
+        /** <inheritdoc /> */
+        public string Label { get; private set; }
+
+        /** <inheritdoc /> */
+        public void Dispose()
+        {
+            try
+            {
+                Close(false);
+            }
+            finally
+            {
+                GC.SuppressFinalize(this);
+            }
+        }
+
+        /// <summary>
+        /// Transaction Id.
+        /// </summary>
+        public int Id
+        {
+            get { return _id; }
+        }
+
+        public ClientSocket Socket
+        {
+            get { return _socket; }
+        }
+
+        /// <summary>
+        /// Returns if transaction is closed.
+        /// </summary>
+        internal bool Closed
+        {
+            get { return _closed; }
+        }
+
+        /// <summary>
+        /// Closes the transaction. 
+        /// </summary>
+        private void Close(bool commit)
+        {
+            if (!_closed)
+            {
+                try
+                {
+                    _ignite.Socket.DoOutInOp<object>(ClientOp.TxEnd,
```

1. `_socket` should be used instead of `_ignite.Socket`
2. Make sure we do not throw exceptions from the finalizer and `Dispose`

---

1. Fixed.
2. What exceptions we shoud cover?

---

I think the only situation to be covered here is a socket disconnect.
We can catch `Exception`, and swallow it only if `_socket` is disconnected, otherwise `throw`.

---

Fixed.